src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -140,37 +331,24 @@ void CoreWorkerDirectTaskSubmitter::OnWorkerIdle(
   // Return the worker if there was an error executing the previous task,
   // the previous task is an actor creation task,
   // there are no more applicable queued tasks, or the lease is expired.
-  if (was_error || current_queue.empty() ||
-      current_time_ms() > lease_entry.lease_expiration_time) {
+  if (!lease_entry.WorkerIsStealing() &&
+      (was_error || current_queue.empty() ||
+       current_time_ms() > lease_entry.lease_expiration_time)) {
     RAY_CHECK(scheduling_key_entry.active_workers.size() >= 1);
 
     // Return the worker only if there are no tasks in flight
     if (lease_entry.tasks_in_flight == 0) {
-      // Decrement the number of active workers consuming tasks from the queue associated
-      // with the current scheduling_key
-      scheduling_key_entry.active_workers.erase(addr);
-      if (scheduling_key_entry.CanDelete()) {
-        // We can safely remove the entry keyed by scheduling_key from the
-        // scheduling_key_entries_ hashmap.
-        scheduling_key_entries_.erase(scheduling_key);
-      }
-
-      auto status =
-          lease_entry.lease_client->ReturnWorker(addr.port, addr.worker_id, was_error);
-      if (!status.ok()) {
-        RAY_LOG(ERROR) << "Error returning worker to raylet: " << status.ToString();
-      }
-      worker_to_lease_entry_.erase(addr);
+      RAY_LOG(DEBUG) << "Number of tasks in flight == 0, calling StealTasksIfNeeded!";
+      StealTasksIfNeeded(addr, was_error, scheduling_key, assigned_resources);
```

Comment:
Actually I don't think that this behavior will match the previous code, right? If there was an error or the lease is expired (`was_error || current_time_ms() > lease_entry.lease_expiration_time`) then we should always return the worker. We should only try to steal tasks if these are not true and the task queue is empty. Can you update this and make sure to add a regression unit test?

Reply:
Just made the changes. What do you mean by _regression_ unit test?

Reply:
Ah sorry I missed this! This means a unit test that breaks with the bug and passes with the fix, so it will catch any future regressions. It's possible there's already one in the current unit tests, but could you check and make sure?